torch/csrc/autograd/engine.cpp

```diff
@@ -727,6 +761,17 @@ auto Engine::execute(const edge_list& roots,
     cb_lock.lock();
   }
 
+  // Syncs leaf streams with default streams (if necessary)
```

Comment:
They could be. I think it's better to do it after:

- Inserting syncs now would cause work on the default stream to wait for backward to be done. If work were being enqueued on the default stream, however, there is no reason for it to wait at this time.
- We would either have to possibly add extra syncs or continue to add the leaf streams to a set and check membership (since the set dedups).